### PR TITLE
Make MkIssueLabels create events when (un)labeling issues

### LINKED ISSUE #1055: MkIssue.events() doesn't track reopenings or previous closures
If you create and then `close()` and then `open()` an `MkIssue`, `MkIssue.events()` doesn't include a `reopened` event.
And if you then re-`close()` that same `MkIssue`, `MkIssue.events()` only includes the latest `closed` event, and not the earlier `closed` event.

### LINKED ISSUE #1056: MkIssueLabels adds/removes don't generate Events
When adding and removing labels from an `MkIssue`, unlike in the real GitHub, no `labeled` or `unlabeled` `Event`s show up in the `MkIssue.events()` stream.

### PR DESCRIPTION
**Pull request by cvrebert**
Fixes #1056.

### CODE REVIEW

**Comment by dmarkov** [comment]
@cvrebert Many thanks for the PR, let me find a reviewer for it

**Comment by dmarkov** [comment]
@carlosmiranda please review, thanks

**Comment by carlosmiranda** [comment]
@cvrebert This branch is too big. Please create separate pull requests for each issue - we avoid combining multiple issues in a single pull requests on this project.

**Comment by cvrebert** [comment]
@carlosmiranda They are basically the same issue, just with different classes, but fine. I have revised this PR and will open a PR for the other commit.
Please now actually review this PR.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java [hunk 1]
```diff
@@ -154,20 +178,31 @@ public Label map(final XML xml) {
...
 
     @Override
     public void clear() throws IOException {
-        this.storage.apply(
-            new Directives().xpath(
-                String.format("%s/label", this.xpath())
-            ).remove()
-        );
+        for (final String label : this.labelSet()) {
+            this.remove(label);
```
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @cvrebert This doesn't do anything - it just removes elements from the set that `labelSet()` returns. Better use `iterator.remove()`.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java [hunk 2]
```diff
@@ -154,20 +178,31 @@ public Label map(final XML xml) {
...
+                this.ticket,
+                this.self,
+                Optional.of(name)
+            );
+        }
     }
 
     @Override
     public void clear() throws IOException {
-        this.storage.apply(
```
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @cvrebert Why did you remove this code? We have a mock XML-based file representation, which this statement modifies.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java [hunk 3]
```diff
@@ -182,4 +217,15 @@ private String xpath() {
         );
     }
 
+    /**
+     * Returns a set of all of the issue's labels.
+     * @return Set of label names
+     */
+    private Set<String> labelSet() {
```
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @cvrebert As far as I can tell, there's no particular reason why it should declare the return type `Set`, let's make the it `Collection` instead. And let's name this method `labels()`, see [our method naming convention here](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#method-names).

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java [hunk 4]
```diff
@@ -154,20 +178,31 @@ public Label map(final XML xml) {
...
+                new Directives().xpath(
+                    String.format("%s/label[.='%s']", this.xpath(), name)
+                ).remove()
+            );
+            final MkIssueEvents events = new MkIssueEvents(
+                this.storage,
+                this.self,
+                this.repo
+            );
+            events.create(
```
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: You can inline this statement from the `new MkIssueEvents()` statement above.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java [hunk 5]
```diff
@@ -154,20 +178,31 @@ public Label map(final XML xml) {
...
-                String.format("%s/label[.='%s']", this.xpath(), name)
-            ).remove()
-        );
+        if (this.labelSet().contains(name)) {
+            this.storage.apply(
+                new Directives().xpath(
+                    String.format("%s/label[.='%s']", this.xpath(), name)
+                ).remove()
+            );
+            final MkIssueEvents events = new MkIssueEvents(
```
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @cvrebert No need to assign it to a variable, just inline the `create` call below.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java [hunk 6]
```diff
@@ -115,11 +119,31 @@ public Issue issue() {
     public void add(@NotNull(message = "labels can't be NULL")
         final Iterable<String> labels
     ) throws IOException {
+        final Set<String> oldlabels = this.labelSet();
```
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @cvrebert We should name this variable `existing`, as per our naming convention [here](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#variable-names).

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java [hunk 7]
```diff
@@ -115,11 +119,31 @@ public Issue issue() {
...
+        final Set<String> addedlabels = new HashSet<String>();
```
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @cvrebert Let's call this variable `added` instead, see our [variable naming convention](http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#variable-names).

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 2 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @carlosmiranda The new code delegates to `MkIssueLabels.remove()`, which does the XML modification and also emits the relevant `unlabeled` events.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 1 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @carlosmiranda This does do things. See https://github.com/jcabi/jcabi-github/pull/1068/files#r28837419

**Comment by carlosmiranda** [comment]
@cvrebert Thanks for the update. Yes, they may be the same problem in two different places, but the fact that you raised two separate issues means that you believe they exist independently and are fixable independently. :) Anyhow, I think it looks good, just some comments to make things inline with the project conventions.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 4 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: Done.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 5 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: Done.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 1 above)
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @cvrebert Duh, silly me, I thought you were calling remove on the collection and not `this`. It's fine don't worry about it anymore, I'm wrong.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 2 above)
Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @cvrebert Makes sense, I didn't see that, my apologies.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 3 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: @carlosmiranda 

> And let's name this method `labels()`

Done.

> As far as I can tell, there's no particular reason why it should declare the return type `Set`

Logically speaking, it is a set; it cannot have duplicates; multiple labels, by definition, cannot share the same name.

Comment (carlosmiranda) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: I get you there, but we aren't using any operations based on the `Set` contract right now, so better to use the most basic interface that we need.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 6 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: Done.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 3 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: Anyway, Done, changed the type.

File: src/main/java/com/jcabi/github/mock/MkIssueLabels.java (hunk 7 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/mock/MkIssueLabels.java]: Done.

**Comment by carlosmiranda** [comment]
@cvrebert Much appreciated, looks great!

**Comment by carlosmiranda** [comment]
@rultor Merge please

**Comment by yegor256** [comment]
@rultor merge

**Comment by cvrebert** [comment]
Thanks!

**Comment by yegor256** [comment]
@cvrebert maybe it's time to release a new version to Maven Central?

**Comment by cvrebert** [comment]
@yegor256 I was going to request that after #1073 gets merged.

**Comment by dmarkov** [comment]
@carlosmiranda done, I added **19 mins** in payment `55947126`, 24 hours and 22 mins spent to complete this. review comments (c=4) added as a bonus. +19 added to your rating, at the moment it is: [+4461](http://www.netbout.com/b/35023?open=rating)

**Comment by dmarkov** [comment]
@rultor please deploy
